codegen/gateway.go

```diff
@@ -206,6 +208,40 @@ func NewHTTPClientSpec(jsonFile string, clientConfigObj map[string]string, h *Pa
 	}, nil
 }
 
+// MiddlewareSpec holds information about each middleware at the endpoint
+// level. The same mid
+type MiddlewareSpec struct {
+	// The middleware package name.
+	Name string
+	// Go import path for the middleware.
+	Path string
+	// Middleware specific configuration options.
+	Options map[string]interface{}
+}
+
+// NewMiddlewareSpec creates a middleware spec from a go file.
+func NewMiddlewareSpec(goFile string) (*MiddlewareSpec, error) {
+	_, err := os.Stat(goFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not find file %s: ", goFile)
+	}
+
+	// Read the go middleware definitions.
+	fset := token.NewFileSet()
+	f, err := parser.ParseFile(fset, goFile, nil, parser.PackageClauseOnly)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err, "Could not read go file %s: ", goFile,
+		)
+	}
+
+	// TODO(sindelar): Consider adding other middleware validation here.
```

Comment:
We might want to have a json module definition file for each middleware instead of parsing the go source code.